fractalis/data/etls/transmart/handler_transmart.py

```diff
@@ -56,9 +57,82 @@ def _get_token_for_credentials(self, server: str, auth: dict) -> str:
+        handle = f'{oidc_server_url}/certs'
+        logger.info('Validating the token...')
+        r = requests.get(handle)
+        if r.status_code != 200:
+            error = "Could not validate the token. " \
+                    "Reason: [{}]: {}".format(r.status_code, r.text)
+            logger.error(error)
+            raise ValueError(error)
+        try:
+            json_response = r.json()
+            key0 = json_response.get('keys')[0]
+            key0_json = json.dumps(key0)
+        except Exception:
+            error = "Could not retrieve the public key. " \
+                    "Got unexpected response: '{}'".format(r.text)
+            logger.error(error)
+            raise ValueError(error)
+        try:
+            public_key = RSAAlgorithm.from_jwk(key0_json)
+        except Exception as e:
+            error = f'Invalid public key!. Reason: {e}'
+            logger.error(error)
+            raise ValueError(error)
+        return key0.get('alg'), public_key
+
+    @staticmethod
+    def get_auth_value(auth: dict, property_name: str) -> str:
+        value = auth.get(property_name, '')
+        if len(value) == 0:
+            raise KeyError(f'The authentication object must contain the non-empty field: "{property_name}"')
+        return value
+
+    def get_client_id_and_url(self, server: str, auth: dict, ) -> (str, str):
+        auth_service_type = self.get_auth_value(auth, 'authServiceType').upper()
+        if auth_service_type == AuthServiceType.OIDC.name:
+            client_id = self.get_auth_value(auth, 'oidcClientId')
+            url = self.get_auth_value(auth, 'oidcServerUrl')
+        elif auth_service_type == AuthServiceType.TRANSMART.name:
+            client_id = 'glowingbear-js'
+            url = server + '/oauth/token'
+        else:
+            raise KeyError("The authentication service type in authentication object has to be one of "
+                           "'oidc', 'transmart'")
+        return client_id, url
+
+    def _get_token_for_credentials(self, server: str, auth: dict) -> str:
+        try:
+            user = self.get_auth_value(auth, 'user')
+            passwd = self.get_auth_value(auth, 'passwd')
+            client_id, url = self.get_client_id_and_url(server, auth)
+        except KeyError as e:
+            logger.error(e)
+            raise ValueError(e)
+        return self.retrieve_token(url, client_id, user, passwd)
+
+    def _validate_request_token(self, server: str, auth: dict):
+        auth_service_type = auth.get('authServiceType', '').upper()
+        if auth_service_type == AuthServiceType.OIDC.name:
+            try:
+                client_id, url = self.get_client_id_and_url(server, auth)
+                algorithm, public_key = self.retrieve_keycloak_public_key_and_algorithm(url)
+                user_token = jwt.decode(auth.get('token'), public_key, algorithms=algorithm, audience=client_id)
```

Comment:
@gijskant this will also throw an exception, when the signature expires. I am wondering, if in this case I should add a way to request another token (using _get_token_for_credentials) or just return 401 error. This happens pretty often, when the token in GB is almost expired.

Reply:
We could also tune Glowing Bear to refresh a token slighly earlier (e.g., when it is less than a minute before expiry)